framework/src/media/audio/audio_manager.c

```diff
@@ -514,43 +519,67 @@ static unsigned int resample_stream_out(audio_card_info_t *card, void *data, uns
 	return resampled_frames;
 }
 
-static audio_manager_result_t get_audio_volume(int fd, audio_config_t *config, audio_io_direction_t direct)
+static audio_manager_result_t get_audio_volume(audio_io_direction_t direct)
 {
+	audio_manager_result_t ret;
+	int fd;
 	struct audio_caps_desc_s caps_desc;
 	uint8_t max_volume;
 	uint8_t cur_volume;
 	pthread_mutex_t *card_mutex;
+	audio_card_info_t *card;
+	audio_config_t *config;
+	char card_path[AUDIO_DEVICE_FULL_PATH_LENGTH];
 
 	caps_desc.caps.ac_len = sizeof(struct audio_caps_s);
 	caps_desc.caps.ac_type = AUDIO_TYPE_FEATURE;
 
+	ret = find_audio_card_if_empty(direct);
+	if (ret != AUDIO_MANAGER_SUCCESS) {
+		return ret;
+	}
+	
 	if (direct == INPUT) {
 		caps_desc.caps.ac_format.hw = AUDIO_FU_INP_GAIN;
+		card = &g_audio_in_cards[g_actual_audio_in_card_id];
 		card_mutex = &g_audio_in_cards[g_actual_audio_in_card_id].card_mutex;
 	} else {
 		caps_desc.caps.ac_format.hw = AUDIO_FU_VOLUME;
+		card = &g_audio_in_cards[g_actual_audio_in_card_id];
 		card_mutex = &g_audio_out_cards[g_actual_audio_out_card_id].card_mutex;
 	}
+	get_card_path(card_path, card->card_id, card->device_id, direct);
 
 	pthread_mutex_lock(card_mutex);
+	fd = open(card_path, O_RDONLY);
+	if (fd < 0) {
+		meddbg("Failed to open device path : %s type : %d errno : %d\n", card_path, direct, errno);
+		ret = AUDIO_MANAGER_DEVICE_FAIL;
+		goto errout;
+	}
+	
 	if (ioctl(fd, AUDIOIOC_GETCAPS, (unsigned long)&caps_desc.caps) < 0) {
 		meddbg("An ioctl error occurs\n");
-		pthread_mutex_unlock(card_mutex);
-		return AUDIO_MANAGER_DEVICE_FAIL;
+		ret = AUDIO_MANAGER_DEVICE_FAIL;
+		goto errout_with_fd;
 	}
 
 	max_volume = caps_desc.caps.ac_controls.hw[0];
 	cur_volume = caps_desc.caps.ac_controls.hw[1];
 
 	/* scale here */
 	cur_volume = cur_volume * AUDIO_DEVICE_MAX_VOLUME / (max_volume - (max_volume % AUDIO_DEVICE_MAX_VOLUME));
-
+	config = &card->config[card->device_id];
 	config->max_volume = max_volume;
 	config->volume = (uint8_t) cur_volume;
-	pthread_mutex_unlock(card_mutex);
 	medvdbg("Max_vol = %d,  cur_vol = %d\n", config->max_volume, config->volume);
 
-	return AUDIO_MANAGER_SUCCESS;
+	ret = AUDIO_MANAGER_SUCCESS;
```

Comment:
How about modifying audio_manager_result_t ret = AUDIO_MANAGER_SUCCESS; in line 524 and deleting this line..?